### ISSUE TITLE
Move PNGs with QR codes out of DB

### ISSUE THREAD

**Issue by siphiuel** [status-im/open-bounty#384]
### Description

*Type*: Feature

PNGs containing QR codes visible in bot comments on GitHub are stored in DB in the `issue_comment` table. Images in bot comments link directly to the OpenBounty server, and so are fetched from DB each time link is accessed.

### Solution
Devise a way of storing PNG binary data outside of DB (either filesystem or external service like IPFS/Swarm/etc.). DB should only store image hashes, or URLs if images will be stored on an external service.

**Comment by commiteth** [comment]
![Contract deploying](http://localhost:3003/img/deploying_contract.png)
Transaction: [0x25a7b837438c021c141d18d922333040795990791b97e4fe30dbecf4e16d7844](https://ropsten.etherscan.io/tx/0x25a7b837438c021c141d18d922333040795990791b97e4fe30dbecf4e16d7844)

**Comment by status-open-bounty** [comment]
![Contract deploying](https://openbounty.status.im/img/deploying_contract.png)
Transaction: [0x25a7b837438c021c141d18d922333040795990791b97e4fe30dbecf4e16d7844](https://etherscan.io/tx/0x25a7b837438c021c141d18d922333040795990791b97e4fe30dbecf4e16d7844)

**Comment by msuess** [comment]
Taking a stab at this :) I'll do a little bit of research first as to which service to use and post my findings here. @siphiuel do you have any preferences or experience regarding that?

**Comment by msuess** [comment]
swarm-gateways.net seems very slow.

I posted some IPFS gateway benchmarks [here](https://github.com/msuess/ipfs-gateway-benchmarks).

**Comment by martinklepsch** [comment]
This is great stuff @msuess! Seems like Infura might be the best option perf-wise? Did any of the other stand out for other reasons?

**Comment by jwagener** [comment]
How about simply rendering the images on the fly?
A CDN could be used in front of the web server, if load / rendering times are an issue.
This would avoid the headache of interacting with external services to store and update the images.

**Comment by martinklepsch** [comment]
@jwagener that's a great idea to consider. Would be interesting to investigate how long one image-rendering takes given that we currently [launch a bunch of processes](https://github.com/status-im/open-bounty/blob/develop/html2png.sh) for it. But yeah I agree, this could be significantly easier :+1: 

@jwagener btw are you in [Riot](https://chat.status.im/#/room/#openbounty:status.im) as well? 

@msuess what do you think about the caching approach?

**Comment by martinklepsch** [comment]
Completely different approach could be using an API like Figma’s: https://blog.figma.com/introducing-figmas-platform-ee681bf861e7

Not even sure if possible but could be a nice opportunity to test drive some of this pretty exciting (imho), new stuff.

**Comment by msuess** [comment]
@jwagener @martinklepsch I think that generating SVGs is a great idea! Maybe we can find a way to do it in-process, without `wkhtmltopdf` (so that's what it's used for, hah). @jwagener did you have any specific caching provider/CDN in mind?

**Comment by martinklepsch** [comment]
Just to double check: if we go with SVG we probably don't need a CDN at all? (for now at least)

**Comment by msuess** [comment]
I'd say probalby not, especially if we can do it in-process.

**Comment by jwagener** [comment]
Had no specific CDN in mind. Also agree that it’s probably not necessary
with the lightweight SVG solution.

things to consider:

- providing a legacy support for the existing embedded Png images. Maybe a
redirect to SVG works, then png support can be dropped completely.

-Embedding the Roboto font.

I guess a good next step is a proof of concept to check if it works fine to
render the current style in SVG.

Do you want to do that? If not I’m happy to jump in.

**Comment by msuess** [comment]
I can look into rendering the QRs as SVGs, if you want :)

I just realized that I'm not sure what we're trying to fix here @martinklepsch @siphiuel. Is this just about reducing the size of the DB or is it about reducing traffic to the OB server?

**Comment by martinklepsch** [comment]
It's about removing image/binary data from DB. Server load is a non-issue at this point :)

**Comment by msuess** [comment]
@denis-sharypin could you please provide the status logo (the on with the "S" in the blue circle) as svg so I can embed it within the QR code SVG? Thank you! :)

**Comment by denis-sharypin** [comment]
@msuess  you can take it here — https://www.dropbox.com/s/djrlvrobuk0u98w/status_logo_blue.svg?dl=0

**Comment by msuess** [comment]
It looks like the embedding of the font might be a bit problematic. I was able to import Roboto as a web font through css, but that only works when embedding as `<svg>`. The CSS won't be evaluated by the browser if the SVG is loaded through an `img` tag. Another solution might be to use SVG Fonts, as described [here](https://stackoverflow.com/questions/11851821/how-i-use-a-custom-font-with-the-batik-svg-library). I'll investigate further.

**Comment by martinklepsch** [comment]
Thanks for keeping us posted @msuess :+1:

**Comment by msuess** [comment]
I embedded the font using a css `@font-face` declaration with a data URI. This works well in Firefox and Chrome, however it blows up the size of the SVG to 400k, which is about 10x the size of the PNGs.

Using some (fontforge)[https://fontforge.github.io/en-US/] magic and my awesome StackOverflow copypaste skills I was able to strip the TTF of all non-ASCII characters with a script lifted from (here)[https://stackoverflow.com/questions/14557944/downsizing-an-otf-font-by-removing-glyphs].

This brings down the size of the output SVG to 28KB, which still only saves us about 6KB.

Keep in mind that this output file still contains some Inkscape stuff that could be stripped, but I don't think it will save that much.

I think the only viable solution at this point would be to programmatically render the SVG, converting the text to paths and stripping the `@font-face` directive. I am still trying to find some Java libraries that can do this.

**Comment by martinklepsch** [comment]
If we can do SVG with 12k gzipped and it works on major browsers that seems reasonable to me 👍 (did I misunderstand anything?)

**Comment by msuess** [comment]
Nope, that's what I said. I'll test in Safari, IE, Mobile as well.

**Comment by martinklepsch** [comment]
Hey @msuess  — hope you had a nice weekend! 🙂  Just wanted to check if there has been any further progress on this? With the GitHub proxy stuff you mentioned in #412 I'm wondering if we could just use GitHub as CDN? 😛 

> Darn. It looks like the github proxy is stripping the `src` attributes from inline SVG CSS definitions. The same thing happens if I host the image somewhere else.

**Comment by georgiemathews** [comment]
@martinklepsch I've started it. let me know if you have any questions

**Comment by endenwer** [comment]
@siphiuel @martinklepsch Can I just save png to filesystem when creating `issue_comment` and store image hash in db. If that's ok I can start working on it.

**Comment by endenwer** [comment]
Hey @andytudhope. Is there anyone who can review my pr?

**Comment by pablanopete** [comment]
Sorry for the long wait @endenwer -- @siphiuel may be able to review but we're putting a resource hold on Open Bounty for now so this bounty would be considered out of commission for the time being. 

cc @andytudhope for revoking
